lib/Sema/SemaExpr.cpp

```diff
@@ -12449,6 +12449,63 @@ ExprResult Sema::CreatePositionalParameterExpr(unsigned Index, QualType QT) {
   return new (Context) PositionalParameterExpr(Index, QT);
 }
 
+ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
+                                     Declarator &D, ParsedType &Ty,
+                                     SourceLocation RParenLoc, Expr *BaseExpr,
+                                     Expr *CountExpr, Expr *RangeExpr,
+                                     BoundsCastExpr::Kind kind) {
+  assert(!D.isInvalidType() && (BaseExpr != nullptr) &&
+         "ActOnBoundsCastExpr(): missing type or expr");
+
+  TypeSourceInfo *castTInfo = GetTypeForDeclaratorCast(D, BaseExpr->getType());
+
+  if (D.isInvalidType())
+    return ExprError();
+
+  ExprResult Res = CorrectDelayedTyposInExpr(BaseExpr);
+  if (!Res.isUsable())
+    return ExprError();
+  BaseExpr = Res.get();
+
+  checkUnusedDeclAttributes(D);
+
+  BoundsExpr *bounds = nullptr;
+  ExprResult Result(true);
+
+  ExprResult ResCount = CorrectDelayedTyposInExpr(CountExpr);
+  ExprResult ResRange = CorrectDelayedTyposInExpr(RangeExpr);
+
+  if (!ResCount.isUsable() && !ResRange.isUsable()) {
+    llvm::APInt I = llvm::APInt(1, 1, false);
+    uint64_t Bits = I.getZExtValue();
+    unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
+    llvm::APInt ResultVal(Width, Bits);
+    IntegerLiteral *One = IntegerLiteral::Create(
+        Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
+    Result =
+        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                             One, SourceLocation());
+  } else if (ResCount.isUsable() && !ResRange.isUsable()) {
+    Result =
+        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                             ResCount.get(), SourceLocation());
+
+  } else if (ResCount.isUsable() && ResRange.isUsable()) {
+    Result = ActOnRangeBoundsExpr(SourceLocation(), ResCount.get(),
+                                  ResRange.get(), SourceLocation());
+  } else {
+    Result = ActOnNullaryBoundsExpr(SourceLocation(), BoundsExpr::Kind::None,
+                                    SourceLocation());
+  }
+  if (Result.isInvalid())
```

Comment:
You should build an invalid bounds expression.   There is a helper method in Sema to do this:  CreateInvalidBoundsExpr().   This will suppress downstream errors in bounds checking etc.